### PR TITLE
Remove exception handling/reporting to the API in favor of logging only

### PR DESCRIPTION
**Pull request by carlosantoniodasilva**
Reporting exceptions through the API was meant to help investigate
issues, but it appears to almost never been useful, and possibly a
source of bugs, so we're removing this handling, and simplifying the
logic to just catch any exceptions and log them with backtraces to the
currently configured logger, so they show up in the client logs.

Having them in the logs will still be helpful to diagnose issues if they
happen, since it will allow clients to grab them as needed. Moreover,
the exception rescuing is necessary due to how Threads are handled in
Ruby: raising inside a thread that is not the main one would by default
terminate it and report its exception to STDERR, but we need the thread
to continue running and reporting scale metrics in this case. So we
catch the exceptions, log them for further analysis if needed, while
keeping the reporter thread up.

### CODE REVIEW

File: lib/judoscale/reporter.rb [hunk 1]
```diff
@@ -87,19 +87,13 @@ def register!(config, worker_adapters)
...
+    def log_exceptions
+      yield
     rescue => ex
-      # An exception was encountered while trying to report the original exception.
-      # Swallow the error so the reporter continues to report.
-      logger.error "Exception reporting error: #{ex.inspect}"
+      # Log the exception but swallow it to keep the thread running and processing reports.
+      # Note: Exceptions in threads other than the main thread will fail silently and terminate it.
+      # https://ruby-doc.org/core-3.1.0/Thread.html#class-Thread-label-Exception+handling
+      logger.error "Reporter error: #{ex.inspect}\n#{ex.backtrace.join("\n")}"
```
Comment (adamlogic) [inline lib/judoscale/reporter.rb:96]: What do you think of logging each backtrace line with a separate call to `logger.error`?

My concern is that I often have customers grep their logs for the "[Judoscale]" tag, and that will miss the backtrace with the current approach.

On the other hand, this really is a single message, so it feels wrong to spread it over multiple logger calls. And gripping for "judoscale" (case-insensitive) instead of "Judoscale" would at least include the backtrace portions that include this gem.

🤔 I think I've talked myself out of this, but still curious for your thoughts.

Comment (carlosantoniodasilva) [reply lib/judoscale/reporter.rb:96]: It would make it easier to search for all the logs involving judoscale for sure, if we were to log each of these individually and they all end up being tagged in the process. I like that search aspect of it, as it can get a bit annoying to have multiple lines of backtrace that don't seem to be tied to anything in particular, except for the first one. (and yeah, this approach wouldn't even tag the first line of the backtrace...)

One possible issue with logging individually is that the lines can be spread over the logs, getting intermixed with other log messages happening from other areas of the app, and show up with slightly different timestamps. It was the reason I avoided two separate `logger.error` calls (one for the exception and the other for the backtrace), so they'd be logged at the exact same moment as a single "unit", even if they're spread across multiple lines, they'd show up "together" one after the other in the logs, with the same timestamp.

But now that you mentioned it, it got me thinking if we could get the best of both worlds: expand the logger to accept an array of messages and have it be responsible to tag each of them, but still log them as a single message in the end.

In other words, something like this:
```ruby
# caller
logger.error ["Reporter error: #{ex.inspect}", *ex.backtrace]
# or maybe just multiple msg args:
logger.error "Reporter error: #{ex.inspect}", *ex.backtrace
```
```ruby
# logger implementation
def error(*msg)
  logger.error msg.map { |m| tag(m) }.join("\n")
end
```

Something like this would allow us to log multiple lines as a single "unity", where each message/line gets tagged to facilitate log searching.

Comment (adamlogic) [reply lib/judoscale/reporter.rb:96]: That's a great idea! 😍

Comment (carlosantoniodasilva) [reply lib/judoscale/reporter.rb:96]: Cool, here's a stab at that implementation: https://github.com/judoscale/judoscale-ruby/pull/14 (pushed as a separate PR to facilitate review, I'll rebase this one on top of that to use it here.)

File: test/reporter_test.rb [hunk 2]
```diff
@@ -7,15 +7,67 @@
...
+        Config.instance.logger = logger
+        stub_request(:post, %r{registrations}).to_return(body: "{}")
+      }
+      after {
+        Reporter.instance.stop!
+      }
+
+      def stub_reporter_loop
+        Reporter.instance.stub(:loop, ->(&blk) { blk.call }) {
+          Reporter.instance.stub(:sleep, true) {
```
Comment (carlosantoniodasilva) [inline test/reporter_test.rb:37]: Another instance where having multiple stubs could help. I might look into it at some point, but didn't want to introduce yet another thing on this one.

File: test/reporter_test.rb [hunk 3]
```diff
@@ -7,15 +7,67 @@
...
+
+    describe "#start!" do
+      let(:string_io) { StringIO.new }
+      let(:logger) { ::Logger.new(string_io) }
+
+      before {
+        # FIXME: even though the config resets after each test, and the logger is set to the config below,
+        # the proxy included in the Reporter class itself (and a couple other places) is shared and doesn't
+        # get reset anywhere. This ensures it gets re-initialized for each test and uses the configured logger.
+        Reporter.instance.instance_variable_set(:@logger, nil)
```
Comment (carlosantoniodasilva) [inline test/reporter_test.rb:20]: Ran across this issue with the logger leaking between tests preventing me from checking the logger output, put this reset here for now but something worth revisiting more broadly.

File: test/reporter_test.rb [hunk 4]
```diff
@@ -7,15 +7,68 @@
...
+        stub_request(:post, %r{registrations}).to_return(body: "{}")
+      }
+      after {
+        Reporter.instance.stop!
+      }
+
+      def run_reporter_start_thread
+        stub_reporter_loop {
+          reporter_thread = Reporter.instance.start!(Config.instance, Store.instance)
+          reporter_thread.join
```
Comment (carlosantoniodasilva) [inline test/reporter_test.rb:31]: Ensures the thread runs fully before we can run the assertions/expectations.

File: test/reporter_test.rb [hunk 5]
```diff
@@ -7,15 +7,68 @@
 
 module Judoscale
   describe Reporter do
-    before { setup_env({"DYNO" => "web.0", "JUDOSCALE_URL" => "http://example.com/api/test-token"}) }
+    before { setup_env({"DYNO" => "web.1", "JUDOSCALE_URL" => "http://example.com/api/test-token"}) }
```
Comment (carlosantoniodasilva) [inline test/reporter_test.rb:10]: Changed to web.1 so it would attempt to collect adapter info. (want to add a test for that specific behavior later)

**Review by carlosantoniodasilva — COMMENTED** [review]
@adamlogic I was able to get some tests for the exception logging here, it required a bit of stubbing around the thread loop but it gets the testing job done.

I think later we can look into pulling out some of the work done inside the thread to a separate method or something that we can test more in isolation without having to worry about these thread / loop / sleep details, but I didn't want to make further changes mixed with these for now. Let me know your thoughts.

**Review by adamlogic — APPROVED** [review]
All good stuff. Thanks!
